pkg/policy/mapstate.go

```diff
@@ -847,15 +997,62 @@ func (ms *mapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapState
 	var (
 		updates, deletes []MapChange
 	)
+	prefixes := getNets(identities, newKey.Identity)
 	if newEntry.IsDeny {
-		ms.ForEachAllow(func(k Key, v MapStateEntry) bool {
-			// Protocols and traffic directions that don't match ensure that the policies
-			// do not interact in anyway.
-			if newKey.TrafficDirection != k.TrafficDirection || !protocolsMatch(newKey, k) {
-				return true
+		// Do not insert if this is a redundant deny.
+		if !ms.denies.Ancestors(newKey, func(k Key, v MapStateEntry) bool {
+			if !v.HasDependent(newKey) && v.HasSameOwners(&newEntry) &&
+				(newKey.Identity == k.Identity ||
+					identityIsSupersetOf(k.Identity, newKey.Identity, ms.denies.getCidrs(k), prefixes)) {
```

Comment:
`identityIsSupersetOf` is never `true` here as `k.Identity == newKey.Identity` due to `ms.denies.Ancestors` only iterating keys `k` that have the same Identity as `newKey`.